### PR TITLE
Improved type hints in documentation

### PR DESCRIPTION
**Pull request by JonathanWenger**
# In a Nutshell
Type aliases are not properly shown in the documentation and usually expanded completely. This can make the documentation very hard to read and the typing information not very useful. 

Screenshot from 2022-03-18 22-23-39

This PR uses Sphinx native functionality to display type hints instead.

# Detailed Description
- Added `probnum.typing` to the API Documentation
- Removed plugin `sphinx-autodoc-typehints`
- Added Sphinx extension `autodoc` 
- Removed colons in some commonly used classes

If a type alias still is expanded completely in the documentation the following should be added to the file:

```
from __future__ import annotations
```

# Other

Note that now the new standard for documenting `Parameters` is to **not** use a colon (:). 
```
Parameters
---------------
arg :
    Argument, where the docs may show no type hint.
betterarg
    Argument which will be documented with a type hint.
```
Otherwise `autodoc` may interpret the type of the parameter as empty.

### CODE REVIEW

**Comment by JonathanWenger** [comment]
Note that currently in some classes type hints for `Parameters` are not shown. This seems to be a bug in `automodapi`. As this might influence the entire way we currently generate the documentation (i.e. switch to a different plugin for api generation) I would prefer to consider this as out of scope for this PR.

File: docs/source/conf.py [hunk 1]
```diff
@@ -54,22 +50,38 @@
...
+    "IntLike": "typing.IntLike",
+    "FloatLike": "typing.FloatLike",
+    "ShapeLike": "typing.ShapeLike",
+    "DTypeLike": "typing.DTypeLike",
+    "ArrayIndicesLike": "typing.ArrayIndicesLike",
+    "ScalarLike": "typing.ScalarLike",
+    "ArrayLike": "typing.ArrayLike",
+    "LinearOperatorLike": "typing.LinearOperatorLike",
+    "NotImplementedType": "typing.NotImplementedType",
+}
```
Comment (marvinpfoertner) [inline docs/source/conf.py:70]:
```suggestion
autodoc_type_aliases = {type_alias: f"typing.{type_alias}" for type_alias in pn.typing.__all__}
```

and then adding the type aliases to `pn.typing.__all__` might scale a little bit better

File: src/probnum/randvars/_constant.py [hunk 2]
```diff
@@ -1,4 +1,5 @@
 """(Almost surely) constant random variables."""
```
Comment (marvinpfoertner) [inline src/probnum/randvars/_constant.py:1]: I did not notice this before, but I'm not sure if we want exactly this comment here. This reads more like "I'm pretty sure that this module contains constant random variables, but we can never know"

**Review by marvinpfoertner — APPROVED** [review]
Minor stuff, otherwise perfect 👍🏻

File: docs/source/conf.py (hunk 1 above)
Comment (JonathanWenger) [reply docs/source/conf.py]: oh yeah this is much better. :+1:

File: src/probnum/randvars/_constant.py (hunk 2 above)
Comment (JonathanWenger) [reply src/probnum/randvars/_constant.py:1]: Changed it to a better description.

File: src/probnum/randprocs/_random_process.py [hunk 3]
```diff
@@ -41,7 +41,7 @@ class RandomProcess(Generic[InputType, OutputType], abc.ABC):
     --------
     ~probnum.randvars.RandomVariable : Random variables.
     ~probnum.randprocs.GaussianProcess : Gaussian processes.
-    ~probnum.randprocs.markov.MarkovProcess : Random processes with the Markov property.
+    ~probnum.randprocs.markov.MarkovProcess : Random processes with the Maqrkov property.
```
Comment (marvinpfoertner) [inline src/probnum/randprocs/_random_process.py:44]:
```suggestion
    ~probnum.randprocs.markov.MarkovProcess : Random processes with the Markov property.
```

**Review by marvinpfoertner — APPROVED** [review]
Just a typo left :)

**Pull request merged**